peer/record.go

```diff
@@ -0,0 +1,188 @@
+
+// ErrPeerIdMismatch is returned when attempting to sign a PeerRecord using a key that
+// does not match the PeerID contained in the record.
+var ErrPeerIdMismatch = errors.New("unable to sign peer record: key does not match record.PeerID")
+
+// PeerRecord contains information that is broadly useful to share with other peers,
+// either through a direct exchange (as in the libp2p identify protocol), or through
+// a Peer Routing provider, such as a DHT.
+//
+// Currently, a PeerRecord contains the public listen addresses for a peer, but this
+// is expected to expand to include other information in the future.
+//
+// PeerRecords are intended to be signed by the peer they describe, and there are no
+// public methods for marshalling and unmarshalling unsigned PeerRecords.
+//
+// To share a PeerRecord, first call Sign to wrap the record in a SignedEnvelope
+// and sign it with the local peer's private key:
+//
+//     rec := NewPeerRecord(myPeerId, myAddrs)
+//     envelope, err := rec.Sign(myPrivateKey)
+//
+// The resulting record.SignedEnvelope can be marshalled to a []byte and shared
+// publicly. As a convenience, the MarshalSigned method will produce the
+// SignedEnvelope and marshal it to a []byte in one go:
+//
+//     rec := NewPeerRecord(myPeerId, myAddrs)
+//     recordBytes, err := rec.MarshalSigned(myPrivateKey)
+//
+// To validate and unmarshal a signed PeerRecord from a remote peer, use the
+// UnmarshalSignedPeerRecord function:
+//
+//    rec, envelope, err := UnmarshalSignedPeerRecord(recordBytes)
+//
+// Note that UnmarshalSignedPeerRecord returns the record as well as the
+// SignedEnvelope that wraps it, so that you can inspect any metadata
+// from the envelope if you need it (for example, the remote peer's public key).
+// If you already have an unmarshalled SignedEnvelope, you can call
+// PeerRecordFromSignedEnvelope instead:
+//
+//   rec, err := PeerRecordFromSignedEnvelope(envelope)
+type PeerRecord struct {
+	// PeerID is the ID of the peer this record pertains to.
+	PeerID ID
+
+	// Addrs contains the public addresses of the peer this record pertains to.
+	Addrs []ma.Multiaddr
+}
+
+// NewPeerRecord creates a PeerRecord with the given ID and addresses.
+// It generates a timestamp-based sequence number.
+func NewPeerRecord(id ID, addrs []ma.Multiaddr) *PeerRecord {
+	return &PeerRecord{
+		PeerID: id,
+		Addrs:  addrs,
+	}
+}
+
+// PeerRecordFromAddrInfo creates a PeerRecord from an AddrInfo struct.
+// It generates a timestamp-based sequence number.
+func PeerRecordFromAddrInfo(info AddrInfo) *PeerRecord {
+	return NewPeerRecord(info.ID, info.Addrs)
+}
+
+// UnmarshalSignedPeerRecord accepts a []byte containing a SignedEnvelope protobuf message.
+// It will try to validate the envelope signature and unmarshal the payload as a PeerRecord.
+// Returns the PeerRecord and the SignedEnvelope if successful.
+func UnmarshalSignedPeerRecord(envelopeBytes []byte) (*PeerRecord, *record.SignedEnvelope, error) {
```

Comment:
So here's what I've been thinking.

1. The `record` package defines an `Record` interface with two methods: `MarshalRecord()` and `UnmarshalRecord()`.
2. Record types such as `PeerRecord` implement that interface.
3. The `record` package exposes a global `RegisterType(mc byte[], typ interface{})` function. This creates a mapping between the multicodec and the type. Packages providing record types call this function in their `init()` func.
4. The `ConsumeEnvelope` function returns an extra out param `interface{}`: the unmarshalled record. We use the envelope's record type and reflection to unmarshal the payload.
5. For the cases where we don't want to use reflection, and we expect a particular type, we can have a `ConsumeTypedEnvelope` function that _takes_ a pointer to the struct to unmarshal the payload into.

Reply:
Ah, I like that!

I keep messing around trying to make something like `Envelope<T>` happen, where `T` is the payload type. I think your proposal is as close as we can get in go :)

Also, I just learned that packages can have an `init` function, so that's cool 😄